Secret-specific protos

cyli opened the pull request:
This does not include any implementation yet - the goal is:

1.  merge #1508 
1.  merge this
1.  open 1-2 secrets-implementation PRs

aaronlehmann commented on api/objects.proto:
```diff
@@ -222,3 +222,34 @@ message Cluster {
+
+	Meta meta = 2 [(gogoproto.nullable) = false];
+
+	map<string, SecretData> secret_data = 3;
```
Please document this field, including the meaning of the map keys.

aaronlehmann commented on api/objects.proto:
```diff
@@ -222,3 +222,34 @@ message Cluster {
+
+	Meta meta = 2 [(gogoproto.nullable) = false];
+
+	SecretSpec spec = 3  [(gogoproto.nullable) = false];
```
Should this type be called `SecretDataSpec`?

aaronlehmann commented on api/objects.proto:
```diff
@@ -222,3 +222,34 @@ message Cluster {
+	// Digest represents the cryptographic digest of the secret data
+	string digest = 4;
+
+	// Size represents the size of the secret
```
How is the size computed, and what units are used?

aaronlehmann commented on api/objects.proto:
```diff
@@ -222,3 +222,34 @@ message Cluster {
+	string digest = 4;
+
+	// Size represents the size of the secret
+	int32 size = 5 [(gogoproto.customname) = "SecretSize"];
```
Should this be `uint32`, or do negative values have a valid meaning?

cyli replied:
Yes, uint32, thank you!

aaronlehmann commented on api/secrets.proto:
```diff
@@ -0,0 +1,77 @@
+	repeated Secret secrets = 1;
+}
+
+// Note CreateSecretRequest does not contain a SecretSpec, since that can contain multiple SecretData objects, which have digests
```
This comment doesn't look right; `CreateSecretRequest` does contain a `SecretSpec`.

aaronlehmann commented on api/secrets.proto:
```diff
@@ -0,0 +1,77 @@
+}
+
+// Note CreateSecretRequest does not contain a SecretSpec, since that can contain multiple SecretData objects, which have digests
+// which should be generated by the
```
sentence cut off?

aaronlehmann commented on api/secrets.proto:
```diff
@@ -0,0 +1,77 @@
+
+message UpdateSecretRequest {
+	SecretSpec spec = 1;
+}
```
I'm a little confused by this. Does this insert a new version? Replace an existing one? How is the new version identified?

cyli replied:
Ah yeah I need to comment this.  It adds a new version, and the new version is identified by the version ID (or as per discussions with @diogomonica maybe just by the digest) on the `SecretData` object.  The plan is for, when this secret is updated, to update the `LatestVersion` field on the `Secret` object with this new version ID.

aaronlehmann replied:
Makes sense, please do comment it. It may make sense to call this something like `AddSecretVersion`, but I suspect other people might prefer to keep the name consistent with the `Update*` methods `controlapi`, so I'm not sure I want to start bikeshedding over this.

aaronlehmann commented on api/objects.proto:
```diff
@@ -222,3 +222,34 @@ message Cluster {
+message SecretData {
+	string id = 1 [(gogoproto.customname) = "ID"];
+
+	Meta meta = 2 [(gogoproto.nullable) = false];
```
I find it a little strange that both the top-level `Secret` object and the `SecretData` object inside it have `Meta` fields. What is the one in `SecretData` used for?

cyli replied:
The one in `SecretData` is mainly for when the secret version was created.  Should I just change this to timestamp instead of Meta, do you think?

aaronlehmann replied:
Yes, I think it should just be a timestamp.

`Meta` is meant to be used in top-level objects and is automatically used by the object store. Right now it contains three fields:

- `Version`: used by the store for object versioning. This presumably wouldn't be used in `SecretData` because `SecretData` values are immutable (right?)
- `CreatedAt`: creation timestamp
- `UpdatedAt`: modification timestamp. Assuming `SecretData` values are immutable, this one wouldn't make sense either.

diogomonica replied:
Just a TS sounds better to me.

aaronlehmann commented on api/types.proto:
```diff
@@ -713,3 +713,33 @@ message ManagerStatus {
+// SecretReference is the linkage between a service and a secret that it uses.
+message SecretReference {
+	// Name is the name of the secret that this reference.
+	string name = 1;
```
Why does this reference the `Secret` object by name instead of by ID?

cyli replied:
The ID is opaque, and presumably people want to get the secret by name?  Should the ID just be the name?

aaronlehmann replied:
I'm not sure.

We have a convention of having an ID in every top-level object, and we probably don't want to break that convention.

But the reason I raised the question is that `SecretDataID` below references an ID. Something feels a little strange about needing both a name and an opaque ID to reference a secret. At least using the ID of the `Secret` and the ID of the `SecretData` would be a little more consistent. But maybe there's a good reason to use a name for the `Secret`; it's hard for me to tell without seeing the rest of the code.

cyli replied:
The `SecretDataID` is the version ID, so it refers to a `Secret`, and the version by `SecretDataID`.  (also, the rest of the code is here:  https://github.com/diogomonica/swarmkit/compare/master...secrets-streaming, although it is a giant blob that we are trying to split out into nicer chunks.  https://github.com/diogomonica/swarmkit/commit/f774f55b9e6141e7de19e024cbbc0d2b26cdc920 specifically does some construction of the `SecretReference` object.)

cyli replied:
Sorry this was collapsed because I fixed a typo in the comment, and it shouldn't have been.  :(

diogomonica replied:
@aaronlehmann ultimately this comes from the fact that the way we want people to interact with secrets is using a `name` and not an ID. Everything else comes from that (using `Name` in `SecretReference` so a human can man to the actual secret on `secrets ls`).

aaronlehmann commented on api/secrets.proto:
```diff
@@ -0,0 +1,77 @@
+
+message UpdateSecretResponse {
+	Secret secret = 1;
+}
```
I notice that `ListSecretsResponse` redacts the acutal secret payload. Is the same true for `UpdateSecretResponse` and `CreateSecretResponse`?

cyli replied:
Yep

cyli commented:
Just to make sure I remembered correctly - did we want the secrets in a different RPC service entirely, because the control one was too big?  Or did we want it in the same RPC service but a different file?  cc @stevvooe

stevvooe commented on api/secrets.proto:
```diff
@@ -0,0 +1,136 @@
+
+// GetSecretRequest is the request to get a `Secret` object given a secret name.
+message GetSecretRequest {
+	string name = 1;
```
Most objects are fetched by id, rather than name.

cyli replied:
I figured you'd want to look up a secret by name vs ID - I guess they should use ListSecret for that?

stevvooe replied:
In that case, you list by name, then use the id to access it via get.

aaronlehmann replied:
> In that case, you list by name, then use the id to access it via get.

Agreed, this is how it works with other objects in `controlapi`.

stevvooe commented on api/secrets.proto:
```diff
@@ -0,0 +1,136 @@
+message ListSecretsRequest {
+	message Filters {
+		repeated string names = 1;
+		repeated string name_prefixes = 2;
```
`ids` and `id_prefixes`, as well.

diogomonica replied:
`id_prefixes` seems useless.

stevvooe replied:
@diogomonica Wrong.

It allows you to implement short-code lookup, so the user can do `abc` instead of `abcdef`.

cyli replied:
For secrets though, the secret IDs aren't necessarily that useful to look up though - mainly people are dealing with the name.  And maybe the type, now that I think about it.  I'll be happy to add `id_prefixes` too, though, for consistency.

stevvooe commented on api/secrets.proto:
```diff
@@ -0,0 +1,136 @@
+// that should be removed.  If no version is provided, the intention is that
+// all versions should be removed.
+message RemoveSecretRequest {
+	// Name is name of the secret to remove.
```
`id`, as well.

cyli replied:
Would this be "remove a secret with this ID, or name if ID is not provided" or "remove a secret with this exact name AND ID"?

stevvooe replied:
We only remove things by `id` and `version`, usually. This has very nice consistency guarantees. Imagine if someone removed `foo` and created a new `foo` in the time that another actor decided to remove `foo`. Using the `id` and `version` makes this operation safe.

aaronlehmann replied:
The `Remove` functions in controlapi don't seem to take a `version` parameter. That might be a nice addition, though.

stevvooe replied:
> The Remove functions in controlapi don't seem to take a version parameter. That might be a nice addition, though.

That sounds kind of like a bug but it is probably okay, as long as it works by id.

stevvooe commented on api/objects.proto:
```diff
@@ -222,3 +222,54 @@ message Cluster {
+	// `SecretData` object and referenced by the `SecretData`'s ID.  This map
+	// stores `SecretData` objects keyed by ID, which lets us reference any
+	// particular version of a secret easily.
+	map<string, SecretData> secret_data = 3;
```
Rather than using a map here, shouldn't these be stored in revision order?

diogomonica replied:
We have timestamps to order them, but the objective here was to have the ability to very quickly retrieve the data for a specific version of a secret without iterating through all the `SecretDatas`

stevvooe replied:
@diogomonica We shouldn't be using the access methodology to decide what the data layout is. When these are serialized, they will be a repeated tuple of (id, secret_data), repeating the id in the serialization. This field should just be a `repeated SecretData secret_data`. The user can always index that in memory, if their access pattern requires it.

stevvooe commented on api/objects.proto:
```diff
@@ -222,3 +222,54 @@ message Cluster {
+	// it for lookups without having to sort the `SecretData` map, although
+	// we will have to sort and get the latest version if the `LatestVersion`
+	// is removed.
+	string latest = 5 [(gogoproto.customname) = "LatestVersion"];
```
`latest_version`.

stevvooe commented on api/types.proto:
```diff
@@ -713,3 +713,33 @@ message ManagerStatus {
+
+// SecretType provides information about what kind of secret this is
+enum SecretType {
+	CONTAINER = 0 [(gogoproto.enumvalue_customname) = "ContainerSecret"];
```
Can we provide some documentation on the meaning of each of these values?

stevvooe commented on api/objects.proto:
```diff
@@ -222,3 +222,54 @@ message Cluster {
+// generated from the secret data in the `spec`, such as the digest and size
+// of the secret data.
+message SecretData {
+	string id = 1 [(gogoproto.customname) = "ID"];
```
Rather than having identifiers for each revision of secret data, it might be better to use the digest and have a linked list to the previous version of the secret data.

diogomonica replied:
We wanted a bit of flexibility managing versions (you can delete a specific version somewhere in the middle). Also, the only problem with using digest is the fact that technically you could have two different versions of a secret with the same digest but with different annotations. We need an ID to identify that particular case.

The other alternative is always overwriting the `SecretData` for a specific digest, but even though it's not a stated goal to have a consistent history of secrets, that would technically remove any old entry with the same content.

stevvooe commented on api/types.proto:
```diff
@@ -713,3 +713,33 @@ message ManagerStatus {
+	// SecretDataID represents the ID of the specific Secret version that we're
+	// referencing. This identifier exists so that SecretReferences don't leak
+	// any information about the secret contents.
+	string secret_data_id = 2 [(gogoproto.customname) = "SecretDataID"];
```
Would it be safer to use the digest here?

aaronlehmann replied:
I suggested using opaque IDs instead of the digest because the digest leaks information about the secret itself.

For example if the digest is `sha256:ef92b778bafe771e89245b89ecbc08a44a4e166c06659911881f383d4473e94f`, it's rather straightforward to find out that the secret is `password123`.

stevvooe replied:
What purpose does that obfuscation serve? In general, we believe that a hash is irreversible, so the target data would only be recoverable if you have the mapping from the digest to the value.

aaronlehmann replied:
> In general, we believe that a hash is irreversible

This assumption is only valid if the data being hashed has very high entropy. It's not generally true for hashes of passwords, which is why compromises of unsalted password hashes very often lead to serious security breaches. I feel like it would be very limiting to put a caveat on Docker secrets that the payloads should never be passwords because the hashes are not treated as confidential. That would be pretty embarrassing.

diogomonica replied:
Technically if the secret has `< 256` bits of information, this would give an attacker a better venue for discovering what the secret is.

Imagine that the secret is a 4 digit PIN. Having the `sha256(PIN)` is still trivially reversible.

Regardless of this, the main issue with using digest is that it doesn't allow us to point to different `SecretDatas` that have the same content (but potentially different annotations).

stevvooe replied:
Alright, forgetting salts, we can drop the digest idea.

Does it make any sense to have these as a linked list? I guess I'm not understand the reasoning behind this extra complexity.

stevvooe commented:
@cyli The structure and file layout of this PR looks optimal.

We may want to discuss with @al but since secrets aren't necessarily a cluster only operation, this will allow us to relocate the component running the actual service. Effectively, we could decouple this from the cluster types, if need be.

al commented:
Not the right @al I'm afraid. Who is this guy? He seems to be rather
prolific. I almost wish I was he.

On 8 Sep 2016 20:23, "Stephen Day" <notifications@github.com> wrote:

@cyli <https://github.com/cyli> The structure and file layout of this PR
looks optimal.

We may want to discuss with @al <https://github.com/al> but since secrets
aren't necessarily a cluster only operation, this will allow us to relocate
the component running the actual service. Effectively, we could decouple
this from the cluster types, if need be.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/docker/swarmkit/pull/1511#issuecomment-245710863>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AAApv97Tp-r9Mb-j6PpcHIISy3mk5gS3ks5qoGC-gaJpZM4J3dfm>
.

stevvooe commented:
@al My bad. Hah.

PTAL @aluzzardi

diogomonica commented on api/types.proto:
```diff
@@ -713,3 +713,33 @@ message ManagerStatus {
+// SecretType provides information about what kind of secret this is
+enum SecretType {
+	CONTAINER = 0 [(gogoproto.enumvalue_customname) = "ContainerSecret"];
+	NODE = 1 [(gogoproto.enumvalue_customname) = "NodeSecret"];
```
@cyli let's remove node for now

sanimej replied:
@diogomonica For the network encryption key we decided to add a new secret type. Should we add a type 'NETWORK' to SecretType ? What is its intended purpose, is it for type checking ? ie: a SecretReference from a Service has to be a secret of type CONTAINER ?

diogomonica replied:
We should add the network specific functionality in a future PR.

stevvooe replied:
@diogomonica We need to think about this use cases now.

We need to cover the registry auth case, as well.

sanimej commented on api/types.proto:
```diff
@@ -713,3 +713,33 @@ message ManagerStatus {
+	// SecretDataID represents the specific Secret version that we're referencing to.
+	// This identifier exists so that SecretReferences don't leak any information
+	// about the secret contents.
+	string secret_data_id = 2 [(gogoproto.customname) = "SecretDataID"];
```
@cyli @diogomonica  Should this be a `repeated` field ? For per network encryption we need three keys. So we need a `SecretReference` in the network object with three versions each corresponding to a key.

diogomonica replied:
You will use three distinct `SecretReference`'s

sanimej replied:
Is there any advantage in doing so ? Since we are referring to the same Secret but different versions it seems natural to list the versions but keep one reference.

diogomonica replied:
`SecretReference` point to a specific version.

stevvooe commented on api/objects.proto:
```diff
@@ -222,3 +222,54 @@ message Cluster {
+// it wraps the `spec` provided by the user with useful information that is
+// generated from the secret data in the `spec`, such as the digest and size
+// of the secret data.
+message SecretData {
```
Major nit: can we call this `SecretValue`? It pluralizes better and gets rid of awkward `SecretData.data` or `for _, data := range datas {` usages. 🐹

cyli commented on api/types.proto:
```diff
@@ -713,3 +713,34 @@ message ManagerStatus {
+enum SecretType {
+	// CONTAINER secrets are user-specified secrets that should be injected
+	// into the container as an environment variable or file.
+	CONTAINER = 0 [(gogoproto.enumvalue_customname) = "ContainerSecret"];
```
To continue a collapsed discussion about what other types there should be, as @stevvooe suggested:

Aside from container secrets, it might be good to include cluster-specific secrets, such as the swarm CA key, so it doesn't have to get written to disk.  Alternately I guess that can live in the cluster, since we probably need to encrypt the raft store at rest once it includes secrets.

What kind of secret does the registry auth one count as?  Are these cluster-wide?  Node-specific?

diogomonica replied:
I don't agree that we need to flesh this out now. We have a `Type` exactly so we can do different things in the future.

/cc @stevvooe

cyli replied:
After some IRL discussion with @diogomonica, we can go with 2 options:

1.  We can have a `NodeInternalSecret` type for things like networks, and registry auth.  Since secrets can have restrictions on them, we can also restrict which nodes have access to which secrets.  Since we're enforcing that user secrets have to be ascii/digits, in order to avoid name conflicts with user-specified secrets, we can require that internal secrets start with 🐳 or something (if the annotations can accept unicode).  And then secrets work the same way for everything.

1.  Another option is to have different concrete types.  So maybe `Secret` can have one of `[]*ContainerSecretValue` or one of `[]*NodeInternalSecretValue`, which has better type safety and would allow us to have different specs or definitions of these different types of secret.  But it would mean a larger number of defined objects and types.

/cc @diogomonica @aaronlehmann @stevvooe

stevvooe replied:
> I don't agree that we need to flesh this out now. We have a Type exactly so we can do different things in the future.

Is this because we want to hit a date or because it doesn't make sense?

If we don't hit the registry auth use case with secrets, then the usefulness of the system is in question. I don't see how we can be putting work into security without resolving these long-standing issues.

I'm sorry to be so annoying here, but I can't really LGTM this until we hit that use case.

stevvooe replied:
After some discussion, it seems we can get rid of the type altogether and replace it with an `internal` bool field. For the most part, we seem to only need to hide this for network secrets. In the future, we could introduce "scope" or "namespace" if we want to add dimensions.

aaronlehmann commented on api/types.proto:
```diff
@@ -624,6 +624,15 @@ message RaftConfig {
+	// MaxSecrets specifies how many user-specified secrets there can be
+	uint32 max_secrets = 1;
+	// MaxSecretVersions specifies how many versions of a secret there can be
+	uint32 max_secret_versions = 2;
```
Will `0` values be interpreted as implying the default values? I'm thinking about the situation where an existing cluster is upgraded to a version that supports secrets. These variables will equal `0` in that case since the cluster object was not created with a `Secrets` config.

cyli replied:
That's a good point, and a great suggestion, thank you.  I was wondering where defaults were set.  That seems good - should I add the default values somewhere here?

aaronlehmann replied:
No, as far as I know protocol buffers doesn't have the concept of a default value. The code that interprets this configuration will have to check for `0` and handle that as a special case.

cyli replied:
Ah ok, thanks.  I'll just update the comment then.

diogomonica commented on api/objects.proto:
```diff
@@ -222,3 +222,47 @@ message Cluster {
 	uint64 encryption_key_lamport_clock = 6;
 }
+
+// Secret represents a conceptual secret that should be passed to a container.
```
`passed to either a container or a node`

diogomonica commented on api/secrets.proto:
```diff
@@ -0,0 +1,133 @@
+	rpc ListSecrets(ListSecretsRequest) returns (ListSecretsResponse) {
+		option (docker.protobuf.plugin.tls_authorization) = { roles: "swarm-manager" };
+	}
+	// CreateSecret creates and return a `CreateSecretResponse` with a `Secret`` based
```
extra ` in Secret

diogomonica commented on api/secrets.proto:
```diff
@@ -0,0 +1,133 @@
+	// - Returns `AlreadyExists` if the `CreateSecretRequest.SecretValueSpec.Annotation.Name`
+	//   conflicts with an existing secret's name.
+	// - Returns an error if the creation fails.
+	rpc CreateSecret(CreateSecretRequest) returns (CreateSecretResponse) {
```
Should we point out that it also fails if the size of the secret is larger than the `MAX_SIZE`?

diogomonica commented on api/secrets.proto:
```diff
@@ -0,0 +1,133 @@
+		option (docker.protobuf.plugin.tls_authorization) = { roles: "swarm-manager" };
+	}
+
+	// UpdateSecret adds the `UpdateSecretRequest.SecretValueSpec` to a `Secret``
```
extra ` in Secret

diogomonica commented on api/secrets.proto:
```diff
@@ -0,0 +1,133 @@
+	// as the latest version, and returns the updated `Secret` in the `UpdateSecretResponse`.
+	// - Returns `NotFound` if no secret with name
+	//   `UpdateSecretRequest.SecretValueSpec.Annotations.Name` is found.
+	// - Returns `InvalidArgument` if the `UpdateSecretRequest.SecretValueSpec` is malformed.
```
Should we point out that it also fails if the size of the secret is larger than the MAX_SIZE?

diogomonica commented on api/secrets.proto:
```diff
@@ -0,0 +1,133 @@
+	//   `UpdateSecretRequest.SecretValueSpec.Annotations.Name` is found.
+	// - Returns `InvalidArgument` if the `UpdateSecretRequest.SecretValueSpec` is malformed.
+	// - Returns an error if the update fails.
+	rpc UpdateSecret(UpdateSecretRequest) returns (UpdateSecretResponse) {
```
Do we use `UpdateSecret` to delete specific `SecretValue`s?

diogomonica commented on api/secrets.proto:
```diff
@@ -0,0 +1,133 @@
+	// - Returns `InvalidArgument` if `RemoveSecretRequest.ID` is empty.
+	// - Returns `NotFound` if the a secret named `RemoveSecretRequest.ID` is not found.
+	// - Returns an error if the deletion fails.
+	rpc RemoveSecret(RemoveSecretRequest) returns (RemoveSecretResponse) {
```
Should this receive `SecretValue` IDs in the case we want to delete a specific version, or will that always be done through `UpdateSecret`?

cyli replied:
Ah that was our previous discussion about whether it'd actually be useful to remove specific versions, or if "update key will push old versions off the list".  I had thought we went with "update key will push old versions off the list", but I may have misunderstood and would also be happy to change it back.  :D

diogomonica replied:
You're absolutely right. This is the right thing.

diogomonica commented:
LGTM

stevvooe commented on api/types.proto:
```diff
@@ -624,6 +624,20 @@ message RaftConfig {
+	// value of 0 will be interpreted as the default maximum value.
+	// If this value is decreased, there will be no change to existing secrets
+	// until the next update, after which extra versions will be culled.
+	uint32 max_secret_versions = 2;
```
revisions?

stevvooe commented on api/types.proto:
```diff
@@ -713,3 +727,34 @@ message ManagerStatus {
+
+	// Mode specifies how this secret should be exposed inside the task.
+	enum Mode {
+		FILE = 0;
```
Let's add a `SYSTEM` mode, which means the secret will be available via direct access, usually at the agent-level, and won't be projected into file or env.

cyli replied:
Just checking - this would set the "internal" bool, correct?

cyli replied:
Or nm, I guess a user can also set "registry auth"

stevvooe replied:
No, internal and system are complementary, but not related. `SYSTEM` would simply mean that it is accessible by "the system" and not projected as env or file.

diogomonica commented on api/objects.proto:
```diff
@@ -239,6 +239,10 @@ message Secret {
 	string name = 4;
+
+	// Whether the secret is an internal secret (not set by a user) or not.
+	// This is just a copy of the data in `SecretValue.Spec.Internal`
```
`SecretValue.Spec` shouldn't have an internal field. A user shouldn't be able to decide wether a secret is internal or not.

cyli replied:
Oops sorry, forgot to remove that comment.  Doing so now.

diogomonica commented on api/secrets.proto:
```diff
@@ -34,29 +32,29 @@ service Secrets {
 	//   or if the secret data is too long or contains invalid characters.
 	// - Returns `AlreadyExists` if the `CreateSecretRequest.SecretValueSpec.Annotation.Name`
 	//   conflicts with an existing secret's name.
+	// - Returns `ResourceExhausted` if there are already the maximum number of allowed
```
I propose we name this error `SetecAstronomy`

stevvooe replied:
Is `SetecAstronomy` a grpc standard error code?

diogomonica replied:
I propose that we send a PR upstream to support it.

stevvooe commented on api/secrets.proto:
```diff
@@ -0,0 +1,136 @@
+// RemoveSecretRequest contains the ID of the secret that should be removed.  This
+// removes all versions of the secret.
+message RemoveSecretRequest {
+	string secret_id = 1 [(gogoproto.customname) = "SecretID"];
```
I saw the conversation above, but is there a way to remove a specific revision of a secret?

diogomonica replied:
No. As of right now, you set a cluster-wide policy of how many revisions you want per secret, and that is it (FIFO queue).

stevvooe commented on api/secrets.proto:
```diff
@@ -0,0 +1,136 @@
+// UpdateSecretRequest specifies a new secret version to add to an existing secret
+// of the same name.
+message UpdateSecretRequest {
+	SecretValueSpec spec = 1;
```
This operation is weird. Should we call this `UpdateSecretValue` to be more accurate?

diogomonica replied:
We're technically updating the `Secret` to contain one more `SecretValue`, and not updating the `SecretValue` itself.

stevvooe replied:
@diogomonica Per our object model, this should take a `SecretSpec`. This doesn't map one to one and is a special case.

cyli replied:
@stevvooe Sorry, I changed the name of `SecretSpec` to be `SecretValueSpec` everywhere, since it is wrapped by a `SecretValue`.  Would you like me to change that back?

diogomonica replied:
I think `SecretValueSpec` here is the right thing, since the spec will be per `SecretValue` and not per `Secret`.

stevvooe replied:
The problem is that this is "UpdateSecret" which implies that we are providing a `SecretSpec`.

Am I making sense here? We have done this everywhere else in the API.

I would expect to have something like this:

```
message UpdateSecretValue {
  string secret_id = 1;
  Version version = 2;
  SecretValueSpec = 3;
}
```

I'd just like to see the API standards followed here. We can write them down to be more clear, but I think we can resolve this without such overhead.

diogomonica replied:
Can we call it `UpdateSecretValue`? Then you have `SecretValue` that includes a `SecretValueSpec`.

stevvooe replied:
@diogomonica I think, but I don't see the reason to have these as embedded objects. It seems like there should be a latest referenced by name, but allow references by ID for earlier versions. Can we do this without the extra level of indirection?

diogomonica replied:
- `Secret`: Object that has a unique name, and aggregates a repeated number of `SecretValue`, that represent multiple versions of that Secret
- `SecretValue`: Object that has a unique ID and the digest of a particular blob of data, associated to a specific `Secret` by virtue of being included in the `Secret`'s `SecretValue` repeated field.
- `SecretSpec`: Object that contains the immutable portions of a secret (stuff that comes from the client, including the name of the secret, annotations and the blob of data.

What you're proposing is collapsing `Secret` and `SecretValue`, having `name` no longer be a unique field, and having a `Secret` include a `SecretSpec` and always attempt to lookup a secret by ID, unless we're trying to update to the latest version, in which case we do a `Secret` search by name, and order the results to use the latest?

diogomonica replied:
What about this:

```
// Secret
message Secret {
	string id = 1 [(gogoproto.customname) = "ID"];
	Meta meta = 2 [(gogoproto.nullable) = false];

       // Name of the secret, so that we can use it easily for lookups. Not unique.
	string name = 3;

       // SecretSpec
	SecretSpec spec = 4  [(gogoproto.nullable) = false];

	// Digest represents the cryptographic digest of the secret data
	string digest = 5;

	// Size represents the size of the secret
	uint32 size = 6 [(gogoproto.customname) = "SecretSize"];

        // Version represents the current version of the secret, so we don't need 
        // to depend on timestamps to order different secrets with the same name.
        uint32 version = 5;
}
```

stevvooe replied:
@diogomonica Is this too big of a change? Will this simplify the API?

cyli replied:
@diogomonica @stevvooe Updated, so we can have a look at what the rest of the protos would look like.

stevvooe commented on api/types.proto:
```diff
@@ -713,3 +727,31 @@ message ManagerStatus {
+	// SecretValueID represents the ID of the specific Secret version that we're
+	// referencing. This identifier exists so that SecretReferences don't leak
+	// any information about the secret contents.
+	string secret_value_id = 2 [(gogoproto.customname) = "SecretValueID"];
```
Do we also need the value of the secret id to ensure this resolves cleanly? For example, if a secret is changed by name, we lose a clean reference here.

stevvooe replied:
Put differently, if the name resolves, but not the secret_value_id, what happens?

diogomonica replied:
That is a good point. However it allows us to actually list a `SecretReference` that has a name that doesn't match the actual value, since under the hood is pointing to an ID of a `Secret` that used to have that name.

At least this way it will always be consistent (the secret with that name no longer exists).

diogomonica replied:
If the name resolves, but not the secret value, we print a warning on the agent stating that the secret the service references is not available.

cyli replied:
In that case this would be an invalid reference.  This is definitely problematic if someone updates the Secret with 10 more versions without updating a service, I think.  @diogomonica ?

cyli replied:
Is there more logic we need to add to secret removal or updating that warns if that would remove a secret that was currently being used?

stevvooe replied:
Ok, so this is a "valid" condition but represents an error. Such tasks should probably fail.

Presumably, we don't allow the deletion of secrets that have an active reference. This will protect against the malicious case.

The analog for the revision would be to not delete an revisions that have an active reference either.

diogomonica replied:
This is why we had an unlimited number of secrets, so we didn't have to deal with this particular issue ;)

Not sure if rejecting an update because it will make a service fail is the right thing to do. Technically people can also create services with the wrong image name, etc etc.

Maybe a warning should be printed if you're about to pop a secret that is in active use?

diogomonica replied:
@stevvooe what is an "active" reference? Technically you can rollback to a previous spec, and the previous secret reference becomes active. When do we decide a reference is no longer in use?

stevvooe replied:
If you know that there is a `SecretReference` referencing a particular value, you cannot delete it without breaking that service or updating the value reference. Since the reference is user input, presumably you can't just update the value unless the user has omitted it.

diogomonica replied:
What if there is a SecretReference referencing a particular value, but in a task in `SHUTDOWN` State?

stevvooe replied:
@diogomonica That needs to be considered but I wouldn't necessarily call that active.

cyli replied:
@diogomonica True :|  Infinite secrets makes this resolving this reference a bit easier, although if we allow a user to delete a secret, we have the same issue (unless we never delete secrets)

cyli replied:
Is there an action to be taken here insofar as this PR?

stevvooe replied:
@cyli Let's make sure we understand this lifecycle and the implications before proceeding. It sounds like we need to define the cleanup behavior when a revision is "actively" reference, and define what constitutes an active reference.

Have we built anything into this model that prevents from changing how these are managed?

diogomonica replied:
I think we should:

- Warn and require a `--force` if a secret is currently being used and is being renamed
- Warn and require a `--force` if a secret is currently being used and is being deleted
- Deal with the inexistence of either a secret name of a secret data gracefully (warn on agent, simply ignore it).

This solves our problems

stevvooe replied:
@diogomonica Why don't we simply disallow delection for secrets that are in use?

diogomonica replied:
Sounds good. What about renames, same thing? Disallow renames of secrets in use, you need to remove it from all services to rename it first?

cyli replied:
Do we allow secret renames?  Currently in our spike implementation if you try to update a secret with a new name it will reject it because you can't change names.

Also, what about updating the secret version?  This can bump older versions off - should we also warn and require a `--force` if the secret is currently being named?

diogomonica replied:
If we follow the new model being discussed on the thread above, we can now deal with specific version deletion and renaming easily:
- If you try renaming a secret whose ID is in use, reject
- If you try removing a secret whose ID is in use, reject

aaronlehmann replied:
Wouldn't the check of whether a secret is in use be quite expensive? I think you'd have to iterate over every task.

diogomonica replied:
@aaronlehmann any alternatives? We could keep a map of secrets in use in memory, and load it on leader election.

aaronlehmann replied:
Yes, maintaining a map like this is possible.

cyli replied:
@diogomonica Should we then remove secret version limits, (e.g. how many versions one can have per name), and instead just have a limit on the total number of secrets?  

In such a scenario, what does `UpdateSecret` do?  If the actual secret data has changed, do we just create a new `Secret` with a new ID?  (e.g. it becomes update or create)

diogomonica replied:
We would remove `UpdateSecret`. Nothing there for you to Update. We now have `CreateSecret` and `RemoveSecret`

cyli replied:
Will `Name` be an annotation on `SecretSpec` anymore?

If we wanted to support renaming, would we `RenameSecret` just replace `UpdateSecret`?  And since a secret wraps a spec, that would need to take a `SecretSpec` too, where the secret data might change?

But if it only existed on `Secret`, then it can be updated without taking a spec.

diogomonica replied:
Rename should always be a user operation, so having it on the `Spec` isn't a big issue.

cyli replied:
Could you elaborate what the renaming would entail?

Does this involve accepting a `SecretSpec` and updating it on an existing secret?  If so, then unless we enforce that the secret data itself does not change, we lose the previous version of that secret data.  Are we going to take the name the user gives us, and synthesize a new `SecretSpec` ourselves with the same data but the new name?

diogomonica replied:
That is a good point. The solution there is to take `Name` out of the `SecretSpec`, but I think we should just not support rename for now.

aaronlehmann commented on api/objects.proto:
```diff
@@ -222,3 +222,35 @@ message Cluster {
+
+	// Name is just the cached name of the secret, also referred to in
+	// `Secret.Spec.Annotations.Name`.
+	string name = 6;
```
If this is the same as the value in `Spec.Annotations.Name` I think it's better to just remove it and rely on `Spec.Annotations.Name`.

cyli replied:
Updated.

aaronlehmann commented on api/objects.proto:
```diff
@@ -222,3 +222,35 @@ message Cluster {
+// and is immutable. It wraps the `spec` provided by the user with useful
+// information that is generated from the secret data in the `spec`, such as
+// the digest and size of the secret data.
+message Secret {
```
In this new model without `SecretValue`, don't you need a way to check which secret with a given name is the newest? I think there may need to be some kind of versioning field (to check which secret is most current, as opposed to the version in Meta, which tracks the last modification to the object). The simplest thing I can think of is an integer on the `Secret` object. When we create a new secret, we'd list all secrets with the same name and set that counter to `max(versions) + 1`.

cyli replied:
Can we sort by the creation timestamp?

aaronlehmann replied:
I prefer not to rely on the timestamps for behavior when possible, because the managers may have clocks that are slightly out of sync, so it would be possible to create a new secret version that is considered older than existing one.

cyli replied:
Makes sense.  Is it possible for two secrets to have the same version ID, if they came in at the same time?  (e.g. both did the calculation)  Would we then tie-break using creation time (which seems ok, since if there was a to add secrets one has to win anyway, and the next secret will have a higher value).

aaronlehmann replied:
It would be possible to write the code to avoid that. A Raft write transaction has an exclusive lock on the Raft store, so if the listing and calculation is done inside the write transaction, a unique version ID can be guaranteed.

cyli replied:
Oh cool :)  Thanks!

cyli replied:
This has since been added :)

cyli commented:
It seems like now we are supporting versioning by allowing the user to create multiple secrets of the same name, but each is actually a separate secret.

In this change, `UpdateSecret` has been removed, because otherwise if the secret data changed that old version would be lost forever.  If we use the Annotation labels to specify node constraints for the secret, they would have to create a new secret with the same name and data.

Alternately, if we wanted to have an `UpdateSecret` api to support changing Annotations we could enforce that the secret data (and the name, if we didn't want to support renaming) cannot change in an `UpdateSecret` call.  Even more alternately we can just let the user lose the previous data if they called the wrong API.

aaronlehmann commented:
> In this change, UpdateSecret has been removed, because otherwise if the secret data changed that old version would be lost forever. If we use the Annotation labels to specify node constraints for the secret, they would have to create a new secret with the same name and data.

> Alternately, if we wanted to have an UpdateSecret api to support changing Annotations we could enforce that the secret data (and the name, if we didn't want to support renaming) cannot change in an UpdateSecret call. Even more alternately we can just let the user lose the previous data if they called the wrong API.

Either alternative sounds reasonable to me. I'm fine with secrets being immutable. But making it possible to change the annotations without allowing changes to the secret data also sounds fine.

diogomonica commented:
I vote for immutability of secrets. Want to change annotations? We create a new secret with the same data.

diogomonica commented:
@stevvooe @al we really want to make progress on this. This is a blocker for Secrets.

al commented:
Would love to help but I'm still not nearly so accomplished as my namesake
so alas I must unsubscribe again :/

On 15 Sep 2016 22:09, "Diogo Mónica" <notifications@github.com> wrote:

> @stevvooe <https://github.com/stevvooe> @al <https://github.com/al> we
> really want to make progress on this. This is a blocker for Secrets.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/docker/swarmkit/pull/1511#issuecomment-247456310>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAApv49inXwsIbSWqVPtriyyVlSO2hPYks5qqbQEgaJpZM4J3dfm>
> .
>

stevvooe commented on api/secrets.proto:
```diff
@@ -0,0 +1,106 @@
+// or all secrets filtered by (id or id prefix or name or name prefix).
+message ListSecretsRequest {
+	message Filters {
+		repeated string secret_ids = 1 [(gogoproto.customname) = "SecretIDs"];
```
`ids`.

stevvooe commented on api/objects.proto:
```diff
@@ -222,3 +222,37 @@ message Cluster {
+	// Secret is created, all other secrets are listed, and Version is
+	// set to max(all other versions) + 1. This is used to determine which
+	// is the latest version of a secret.
+	uint64 version = 7;
```
Can `Meta.Version` handle this?

stevvooe replied:
That or we define this as a lamport clock.

stevvooe replied:
So, per conversation with @aaronlehmann, defining this as lamport clock is the right model. Similar to network encryption keys.

cyli replied:
updated the name and comment, as per the lamport fields in tasks.

diogomonica commented:
Crap. Sorry @al. 

@aluzzardi you *need* to change your slack username to be aluzzardi. Writting @al in one place and @aluzzardi in another is killing me.

Alternatively, you could buy al off of @al. I'm pretty sure he would be reasonable on the price.

al commented:
I assure you fine ppl that I will be totally unreasonable in my demands but
still let's pursue this ...

On 15 Sep 2016 22:17, "Diogo Mónica" <notifications@github.com> wrote:

> Crap. Sorry @al <https://github.com/al>.
>
> @aluzzardi <https://github.com/aluzzardi> you *need* to change your slack
> username to be aluzzardi. Writting @al <https://github.com/al> in one
> place and @aluzzardi <https://github.com/aluzzardi> in another is killing
> me.
>
> Alternatively, you could buy al off of @al <https://github.com/al>. I'm
> pretty sure he would be reasonable on the price.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/docker/swarmkit/pull/1511#issuecomment-247458128>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAApv6e15ugANwijAVENw9IuaoKgJkzFks5qqbXwgaJpZM4J3dfm>
> .
>

aaronlehmann commented on api/secrets.proto:
```diff
@@ -0,0 +1,106 @@
+// or all secrets filtered by (id or id prefix or name or name prefix).
+message ListSecretsRequest {
+	message Filters {
+		repeated string ids = 1 [(gogoproto.customname) = "SecretIDs"];
```
I'd prefer the `customname` to match the field name in the protobuf source.

cyli replied:
updated

aluzzardi commented on api/dispatcher.proto:
```diff
@@ -202,4 +202,14 @@ message AssignmentsMessage {
+	// UpdateSecrets is a set of new or updated secrets for this node.
+	// In the first assignments message, it contains all of the secrets
+	// the node needs for itself and its assigned tasks.
+	repeated Secret update_secrets = 6;
```
Too late now, but we should have used one level of indirection.

Assignments should be part of the data model - **not** the RPC protocol as it is now.

What I mean by that is we should have had an `Assignment` object type with a `oneof`. The dispatcher protocol therefore (`AssignmentsMessage`) should have been using that instead (rather than embedding a pseudo definition of assignments inline).

With this approach, adding a new type of assignment requires making changes to the dispatcher RPC which is broken.

This is not novel - that's what we're already doing for Raft:
https://github.com/docker/swarmkit/blob/master/api/raft.proto#L119

/cc @stevvooe @aaronlehmann

stevvooe replied:
@aaronlehmann @aluzzardi I agree but that is outside the scope of this PR. Let's open an issue for this.

aluzzardi replied:
Proposal:

```proto
message Assignment {
	oneof item {
		Task task = 2;
		Secret secret = 3;
	}
}

message AssignmentChange {
	enum AssignmentAction {
		UPDATE = 1 [(gogoproto.enumvalue_customname) = "AssignmentActionUpdate"];
		REMOVE = 2 [(gogoproto.enumvalue_customname) = "AssignmentActionRemove"];
	}

	Assignment assignment = 1;
	AssignmentAction action = 2;
}

message AssignmentsMessage {
 	enum Type {
 		COMPLETE = 0;
 		INCREMENTAL = 1;
 	}
 
 	Type type = 1;
 
 	string applies_to = 2;
 	string results_in = 3;

	 repeated AssignmentChange changes = 4;
}
```

aluzzardi commented on api/objects.proto:
```diff
@@ -222,3 +222,34 @@ message Cluster {
+	uint32 size = 5 [(gogoproto.customname) = "SecretSize"];
+
+	// Whether the secret is an internal secret (not set by a user) or not.
+	bool internal = 6;
```
Why does it matter?

cyli replied:
We may want to use secrets management to manage network keys, which will be set internally when networks are created.

We wanted users to not be able to list/remove these secrets that shouldn't necessarily be within their control, so this is a way of distinguishing which secrets should be within the user's purview and which should not.

aluzzardi commented on api/objects.proto:
```diff
@@ -222,3 +222,34 @@ message Cluster {
+	bool internal = 6;
+
+	// Timestamp from the lamport clock to identify the relative age of the secret.
+	uint64 lamport_time = 7;
```
Why can't we use `Meta.Version` instead?

cyli replied:
I might be misunderstanding how `Meta.Version` works, but I think the version is only incremented if that specific `Secret` object is updated?  We wanted a global ordering of the `Secret` objects, so we can tell which one is the latest version of a particular name.

aluzzardi commented on api/secrets.proto:
```diff
@@ -0,0 +1,106 @@
+syntax = "proto3";
```
As I mentioned to @stevvooe, I feel strongly about having this as part of the control API.

We have a very simple concept in place since day 1: Cluster state can only be altered through the Control API - that's it.

cyli replied:
As suggested by @stevvooe, it'd probably be good to list the pros and cons of having a single control API for everything, vs separate control APIs for each type of subsystem:

Everything in 1 monolithic control API protocol:
1.  Pro: Organization is simple, and anyone looking at this one protocol can see what can or cannot be changed about the cluster
1.  Pro: It's easy to see and enforce consistency across function arguments and naming

Control APIs split up across subsystems:
1.  Pro:  Downstream libraries can more tightly specify their dependencies: for example, they may need only a subset of behavior rather than entire control API.  For example, maybe they may depend only on secrets, and could be provided with something that provides the the same secrets API rather than all the whole control API (e.g. providing some sort of secrets management for your build time secrets on your build machine).
1.  Pro:  As @stevvooe suggests it's easier to version the control APIs for different components.  So if the API needed to change for a single component, we could have 2 different version implementations for just that component, as opposed to the whole control API.
1.  Pro: As @stevvooe suggests, it may allow users to switch to different implementations of the different subcomponents?  (e.g. volumes or networks, or even secret backends).
1.  Con: None of this actually exists yet, so do we want to do this now as opposed to deal with splitting it up later?

(I've probably missed some, just jotted some stuff down from convo earlier)

Are we ever going to want to support more modularity in terms of the control of components?  (e.g. differentially versioning certain components, swapping out implementations, providing partial implementations?)  If so, maybe it's worthwhile to start splitting some stuff up now?

cyli replied:
For secrets, at least, we would like the option to have different backends (such as vault or whatever someone is already using for secrets management).  Being able to swap that or have third parties implement a pluggable backends would be very useful.

aluzzardi commented on api/specs.proto:
```diff
@@ -270,4 +275,19 @@ message ClusterSpec {
+	bytes data = 2;
+
+	// When this secret-spec should expire and no longer be available
+	Timestamp expiry = 3;
```
Use case?

cyli replied:
If a user wanted an auto-expiring key that shouldn't be available after a certain period, for instance if the secret is the key to a cert that expires, they may want to ensure that services can no longer be started with said key.

aluzzardi commented on api/types.proto:
```diff
@@ -624,6 +624,14 @@ message RaftConfig {
+message SecretsConfig {
+	// MaxSecrets specifies how many user-specified secrets there can be.  A value
+	// of 0 will be interpreted as the default maximum value.
+	uint32 max_secrets = 1;
```
?

cyli replied:
oops, sorry, typo.

cyli replied:
Oh, wait, spoke too soon.  The proto indexes do start at 1.  Could you elaborate on your question?

cyli replied:
Oh, wait, spoke too soon. The proto indexes do start at 1. Could you elaborate on your question?

aluzzardi commented:
> @aluzzardi you need to change your slack username to be aluzzardi. Writting @al in one place and @aluzzardi in another is killing me.

Of course @diogo.

Woops, I meant @diogomonica, sorry @diogo :)

(and sorry @al too)

cyli commented:
Hey all, thanks for the discussion.  :)  Have pushed a rebased version with the following changes as discussed:

1.  Removed expiry field in the secret spec (expiring secrets as a feature has been explicitly requested by potential users, though, so it might be nice to support this at some point in the future)
1. Removed secret.proto and put the operations on the control API.
1. Left the max number of secrets so that we can limit further creation if we needed
1. Left the global secret lamport clock because since users have to version their own via name, it'd be nice to have a reliable ordering when they list

aaronlehmann commented:
Thanks for the really quick turnaround!

> Left the global secret lamport clock because since users have to version their own via name, it'd be nice to have a reliable ordering when they list

I know I was the one who asked for this, but I feel it may not be worth the complexity anymore now that there's no internal need for it. For user display purposes, I think ordering by timestamp would work well enough.

aaronlehmann commented on api/types.proto:
```diff
@@ -747,3 +755,31 @@ message ManagerStatus {
+
+// SecretReference is the linkage between a service and a secret that it uses.
+message SecretReference {
+	// Name is the name of the secret that this references.
```
As discussed, this name will be informational, and `secret_id` below will be used to actually resolve the secret. Let's make this clear in the comments.

cyli replied:
Good point.  I've also moved this to the end of the message (although I don't know that ordering implies any sort of importance).

sanimej commented:
@cyli @aanand Did we decide to drop the lamport_time field ? For network gossip encryption we need a reliable ordering to delete the old key and insert the new key on every rotation. It can be handled in the client side by having a lamport_time for each secret reference. But if its part of the Secret it will be convenient.

cyli commented:
@sanimej I've dropped it as per @aaronlehmann's suggestion, but I can also easily add it back if it's needed.

aaronlehmann commented:
How are network gossip keys going to be managed in a world where we have immutable secrets and no built-in versioning?

sanimej commented:
`How are network gossip keys going to be managed in a world where we have immutable secrets and no built-in versioning?`

For each network we will have a SecretSpec and three SecretReference in the Network object corresponding to three keys needed for gossip encryption. Network key manager in the swarm leader will be responsible for deleting the old secret, create a new one on each rotation and update the reference in the Network object. Having a lamport_time will avoid having to maintain extra state in the client (key manager in this case).

Updated network object will be sent in the Assignment stream only to the nodes that have task with that network attachment.

sanimej commented on api/control.proto:
```diff
@@ -293,3 +331,59 @@ message UpdateClusterRequest {
+// removes all versions of the secret.
+message RemoveSecretRequest {
+	string secret_id = 1 [(gogoproto.customname) = "SecretID"];
+}
```
Can we make this a `repeated` field ?

cyli replied:
I have no objection, but none of the other remove requests work this way, so would like further input from @aaronlehmann or @stevvooe or @aluzzardi  :)

cyli replied:
Sorry, to clarify this statement I think supporting removing multiple objects in a single transaction might make sense, especially since the network secrets come bundled together.  However, it would be an inconsistency in what the Remove argument takes.

I think it wouldn't be too hard to support this for other objects in the control API as well on the server side, but migrating their existing APIs over to support that may be tricky, if previously it took a single string and now it takes a list of strings.

Will the network manager actually be using the control API, or will it be updating the secret store behind the scenes?  There is also an `internal` bool on the `Secret` object, which the control API does not provide any way to set - I think the assumption was that the network manager would create that `Secret` itself and store it?

aaronlehmann replied:
It's behind the scenes, so it can use a transaction.

I think having `RemoveSecret` remove a single secret is most consistent with the other external APIs.

cyli replied:
Sounds good.  Leaving this as is, then, and have also removed the max number of secrets in the cluster.

aluzzardi replied:
Agreed - consistency first.

It makes sense, but if we want to do it, we should do it across the board on every other `Remove*` request so it's outside of the scope of this PR.

aaronlehmann commented:
Thanks for explaining.

I think it may be preferable to keep the state in the key manager, versus maintaining the `lamport_time` field.

The reason is that `lamport_time` is inefficient to maintain. It involves listing all secrets every time a new secret is created, or maintaining state somewhere else to keep track of this. I expect the key manager would have less overhead because it doesn't need to maintain a global ordering of all secrets (maybe it could just keep the `SecretsReference`s sorted inside the `Network` object)?

sanimej commented:
`I think it may be preferable to keep the state in the key manager, versus maintaining the lamport_time field.`

If we don't foresee other users/clients of the secret infra having a similar requirement its probably better to leave the versioning part to the key manager.

stevvooe commented on api/types.proto:
```diff
@@ -658,6 +658,7 @@ message RaftConfig {
 	uint32 election_tick = 5;
 }
 
+
```
Extra space. ಠ_ಠ

cyli replied:
Removed

stevvooe commented on api/types.proto:
```diff
@@ -747,3 +748,32 @@ message ManagerStatus {
+
+	// Name is the name of the secret that this references, but this is just provided for
+	// lookup/display purposes.  The secret in the reference will be identified by its ID.
+	string name = 4;
```
SecretName?

cyli replied:
Done

stevvooe commented:
LGTM

aluzzardi commented on api/specs.proto:
```diff
@@ -271,3 +275,11 @@ message ClusterSpec {
+	Annotations annotations = 1 [(gogoproto.nullable) = false];
+
+	// Secret payload.
+	bytes data = 2;
```
nit: naming? Not opposed to `data` but wondering if someone has a better name in mind (@stevvooe ?)

nit: Add a comment to data to explain the maximum allowed length

stevvooe replied:
This is as good as anything.

aluzzardi replied:
`data` it is - discard that comment

cyli replied:
Included a max size constant in the `secrets.go` control API stub, and added a comment here.
